pyccel/ast/operators.py

```diff
@@ -796,15 +867,18 @@ class PyccelBooleanOperator(PyccelOperator):
     arg2: PyccelAstNode
         The second argument passed to the operator
     """
-    __slots__ = ()
-    _default_precision = default_precision['bool']
+    __slots__ = ('_dtype','_precision','_shape','_rank',)
```

Comment:
The `_dtype`, `_precision`, `_shape`, and `_rank` don't need to be slots. They are always equal to `NativeBool()`, `default_precision['bool']`, `()`, and `0` so these should be class properties. The setters should then be overridden with:
```
def _set_dtype(self):
    pass
def _set_shape_rank(self):
    pass
```

Reply:
I tried doing this in [6d59098](https://github.com/pyccel/pyccel/pull/797/commits/6d59098c7b9b632c977a4c8d7d5321e3fe7fc58e) but it generated some problems so i retrieved it in [9cd1fc1](https://github.com/pyccel/pyccel/pull/797/commits/9cd1fc168b74f0f27d9bfc0be5dc22f93bb61f0e)

Reply:
The reason it didn't work in 6d59098 is because you didn't set the class property. You need to make sure that `self._dtype` returns something for each class. Either it will return the class property (same object for every instance) or it will return the attribute set in `_set_dtype`. It looks like in 6d59098 you removed it from `_set_dtype` but hadn't yet added it as a class property. Hence the error (I imagine it failed in one of the methods of `PyccelAstNode` which tries to return this property)